### ISSUE TITLE
Seceded tasks don't release resources if they fail

### ISSUE THREAD

**Issue by crusaderky** [dask/distributed#6565]
The long-running -> error worker transition does not release worker resources.

On a related note, there's an unused method `transition_long_running_rescheduled` that should be cleaned up.

**Comment by crusaderky** [comment]
Also, there's nothing that ever removes keys from the `long_running` set.

**Comment by RaphaelRobidas** [comment]
I launch tasks from tasks and secede in the initial tasks while their subtasks complete. It seems like the resources of the initial tasks are not released, which can lead to a situation where no work is getting done since every thread is waiting for resources to be freed up. This issue is quite similar to yours.

**Comment by fjetter** [comment]
xref https://github.com/dask/distributed/issues/5882 this is not directly related but also connected to the handling of long-running state

**Comment by crusaderky** [comment]
@RaphaelRobidas to my knowledge resources are released upon *successful* completion. Do you have a reproducer that shows otherwise?

**Comment by RaphaelRobidas** [comment]
@crusaderky yes, but seceding during the task is not successful completion and does not release resources. In my understanding, seceding is meant to free up the current worker and allow it to execute other tasks. Logically, the resources should be temporarily released as well.

See the example below: the jobs will never complete and the whole cluster will get stuck doing nothing. Is this a related issue or am I using resources incorrectly?

```
import time

import dask
from dask.distributed import (
    Client,
    LocalCluster,
    get_client,
    secede,
    rejoin,
    as_completed,
)


def subjob(arg):
    print("Starting subjob")

    time.sleep(5)

    print("Subjob done")
    return 0


def job(arg):
    print("Starting job")

    time.sleep(1)  # Wait until all jobs are submitted
    client = get_client()
    futs = [client.submit(subjob, i, resources={"CPU": 1}) for i in range(5)]

    secede()
    rets = client.gather(futs)
    rejoin()

    time.sleep(5)
    print("Job done")

    return 0


if __name__ == "__main__":
    with dask.config.set({"distributed.worker.resources.CPU": 8}):
        cluster = LocalCluster(n_workers=1, processes=False)

        client = Client(cluster)

        futures = [
            client.submit(
                job,
                i,
                resources={"CPU": 1},
            )
            for i in range(16)
        ]

        rets = client.gather(futures)

        print("This will never be reached")
```

**Comment by RaphaelRobidas** [comment]
My workaround ended up not working very well at all. I need the ability to release resources when seceding for my uses, so I dug a bit in the code. I found that it is possible to manually release and reattribute resources when seceding/rejoining:
```
def job():
    [...]
    secede()
    with release_resources(CPU=1):
        res = client.gather(futures)
    rejoin()
```
I have defined the function `release_resources`, which modifies the `available_resources` of the current worker. The context exit blocks until the resources can be reserved for the current thread again. This is fine for my usage and works as expected, as far as I can tell. 

Is there an obvious problem with this approach and is this something that would be worth merging into the main branch? I can prepare a PR so you can have a look.

**Comment by crusaderky** [comment]
Please _never, ever_ bypass `Worker.handle_stimulus` and modify the WorkerState directly.

**Comment by RaphaelRobidas** [comment]
Thanks for the help, I adjusted the code to make use of `Worker.handle_stimulus`. I will submit a PR for those who might also benefit from it.
